### PR TITLE
Outputs Update #2 - Run Gudrun in Temporary Directory

### PR DESCRIPTION
**Pull request by rhinoella**
Stage 2 of Outputs Update #457, builds off of #458.

This PR changes the way gudrun_dcs is run, by running it in a temporary directory, so all the files are outputted there, and then organising and copying the outputs back in the user's permanent directory.

This approach has three advantages:
  - Removes the need for dangerous file-management operations (folder/file deletions) in the permanent directory
  - If the program crashes or there is an issue with the run there will be no unorganised files to manage
  - Much more control on how to present outputs to users and choosing save locations

## Changes
- Gudrun and Purge DCS are run in a temporary directory
- The running of the process is done using a QThread and a worker instead of a QProcess as this allows for the use of a context manager
- Purge dialog is changed - no longer asks to run with default parameters (the user can just click 'OK' for this)

## Next Steps:
- Update Batch Processing to utilise the new system- remove `dcs()` function
- Deploy GudPy class to handle all the GudPy processes (remove responsibility from GudrunFile class)
- Add new functionaity to headless mode
- Import project from project folder

### CODE REVIEW

**Comment by rhinoella** [comment]
Still need to:

- [x] Have output tree clearly show 'Default' run
- [x] Implement changes on 'iterate by composition'
- [x] Fix progress bar

File: gudpy/core/gudrun_file.py [hunk 1]
```diff
@@ -145,33 +146,40 @@ def __init__(self, path=None, format=Format.YAML, config_=False):
             If a new input file should be constructed from a config
         """
 
-        self.path = path
-        self.filename = os.path.basename(path)
+        # Path and filename of initial input file
```
Comment (rhinoella) [inline gudpy/core/gudrun_file.py]:
```suggestion
```

File: gudpy/core/gudrun_file.py [hunk 2]
```diff
@@ -270,15 +290,11 @@ def parseInstrument(self):
         """
         try:
             # Initialise instrument attribute to a new instance of Instrument.
```
Comment (rhinoella) [inline gudpy/core/gudrun_file.py]:
```suggestion
```

File: gudpy/core/gudrun_file.py [hunk 3]
```diff
@@ -1461,15 +1477,18 @@ def __str__(self):
 
     def save(self, path='', format=None):
 
+        print(f"saving to {path}")
+
```
Comment (rhinoella) [inline gudpy/core/gudrun_file.py]:
```suggestion
```

File: gudpy/core/gudrun_file.py [hunk 4]
```diff
@@ -145,33 +146,39 @@ def __init__(self, path=None, format=Format.YAML, config_=False):
             If a new input file should be constructed from a config
         """
 
-        self.path = path
-        self.filename = os.path.basename(path)
+
```
Comment (rhinoella) [inline gudpy/core/gudrun_file.py]:
```suggestion
```

File: .github/workflows/test/action.yml [hunk 5]
```diff
@@ -1,5 +1,8 @@
 name: Test
 
+env:
+  ACTIONS_STEP_DEBUG: true
+
```
Comment (trisyoungs) [inline .github/workflows/test/action.yml]: Not used anywhere?

File: gudpy/test/test_gudpy_yaml.py [hunk 6]
```diff
@@ -13,6 +13,9 @@ def testYAML(self):
         gf1.write_yaml("test/TestData/NIMROD-water/water.yaml")
         gf2 = GudrunFile("test/TestData/NIMROD-water/water.yaml")
 
+        print(gf1.instrument.__dict__)
+        print(gf2.instrument.__dict__)
+
```
Comment (trisyoungs) [inline gudpy/test/test_gudpy_yaml.py]: Debugging output?

**Review by trisyoungs — APPROVED** [review]
As much as I'm able to go through such a large PR and comprehend the sweeping changes that it enforces, this looks pretty good! :smiley:

**Comment by trisyoungs** [comment]
Apart from the failing tests, of course....

File: gudpy/test/test_gudpy_yaml.py (hunk 6 above)
Comment (rhinoella) [reply gudpy/test/test_gudpy_yaml.py]: This was to fix the tests! Will be removed :)

**Pull request merged**